File: src/eqn/ast/EqnAstNodePreDefinedFunction.java
package eqn.ast;

import eqn.parser.exception.EqnException;
import org.jetbrains.annotations.NotNull;

import java.util.Vector;

public abstract class EqnAstNodePreDefinedFunction extends EqnAstNode {


    public EqnAstNodePreDefinedFunction(String value) {
        super(value, Type.PreDefinedFunction, PrecedenceType.Function);
        operands = new Vector<>(0);
    }

    protected int arity() {
        return operands.size();
    }

    public void addOperand(@NotNull EqnAstNode operand) throws EqnException {
        this.operands.add(operand);
    }

    @Override
    public String toString() {
        StringBuffer stringBuffer = new StringBuffer(value + '(');
        for (int i = 0; i < operands.size() - 1; ++i) {
            stringBuffer.append(operands.elementAt(i).toString() + ", ");
        }
        stringBuffer.append(operands.lastElement().toString());

        stringBuffer.append(")");
        return new String(stringBuffer);
    }

}

File: src/eqn/ast/EqnAstNodePreDefinedFunction.kt
package eqn.ast

import eqn.ast.base.EqnAstNode
import eqn.ast.base.EqnAstNodeArbitraryArity
import eqn.parser.exception.EqnException

abstract class EqnAstNodePreDefinedFunction(value: String) : EqnAstNodeArbitraryArity(value, Type.PreDefinedFunction, PrecedenceType.Function) {
    override fun arity(): Int {
        return operands.size
    }

    @Throws(EqnException::class)
    override fun addOperand(operand: EqnAstNode) {
        operands.add(operand)
    }

    override fun toString(): String {
        val stringBuffer = StringBuffer("$value(")
        for (i in 0 until operands.size - 1) {
            stringBuffer.append(operands.elementAt(i).toString() + ", ")
        }
        stringBuffer.append(operands.last().toString())
        stringBuffer.append(")")
        return String(stringBuffer)
    }
}